### PR TITLE
refactor(semantic): check duplicate parameters in Binder of FormalParameters

### PR DESCRIPTION
**Pull request by Dunqing**
`check_duplicate_bound_names` will hurt performance

### CODE REVIEW

**Comment by Dunqing** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1840?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #1792** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1792?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #1793** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1793?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #1794** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1794?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #1820** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1820?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #1826** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1826?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #1827** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1827?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #1838** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1838?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #1839** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1839?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #1840** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1840?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/1840?utm_source=stack-comment).

**Comment by Dunqing** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #1838** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1838?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #1846** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1846?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #1840** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1840?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #1839** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1839?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/1840?utm_source=stack-comment).

**Review by IWANABETHATGUY — DISMISSED** [review]

File: crates/oxc_semantic/src/binder.rs [hunk 1]
```diff
@@ -182,9 +185,26 @@ impl<'a> Binder for Function<'a> {
...
+        // Check if duplicate parameters are not allowed
+        let is_not_allowed_duplicate_parameters =
+            // 1. Duplicate parameters not allowed in TypeScript
+            builder.source_type.is_typescript() ||
+            // 2. Duplicate parameters not allowed in strict mode
+            builder.strict_mode() ||
+            // 3. Duplicate parameters not allowed if the function is an arrow function
+            builder.scope.get_flags(builder.current_scope_id).is_arrow_function() ||
+            // 3. Duplicate parameters not allowed if the parameter is not a simple parameter list
+            !self.is_simple_parameter_list();
```
Comment (Boshen) [inline crates/oxc_semantic/src/binder.rs:201]: These conditions don't look right to me, let's discuss it when I'm free.

File: crates/oxc_linter/src/rules/unicorn/no_array_for_each.rs [hunk 2]
```diff
@@ -114,12 +114,12 @@ fn test() {
         r"1?.forEach((a, b) => call(a, b))",
         r"array.forEach((arrayInArray) => arrayInArray.forEach(element => bar(element)));",
         r"array.forEach((arrayInArray) => arrayInArray?.forEach(element => bar(element)));",
-        r"array.forEach((element, index = element) => {})",
+        r"array.forEach((element, index = element1) => {})",
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/unicorn/no_array_for_each.rs:117]: This shouldn't trigger a duplicate parameter error, shall we revert the change?

Comment (Dunqing) [reply crates/oxc_linter/src/rules/unicorn/no_array_for_each.rs:117]: Yes, we need to revert
